### PR TITLE
Programmatic switch between canvas renderer and WebGL

### PR DESCRIPTION
**Pull request by ankusht-work**
Addresses: https://github.com/theia-ide/theia-trace-extension/issues/122

### CODE REVIEW

**Review by bhufmann — APPROVED** [review]
Looks good to me. I've tested both webgl2 and canvas creation by using different environments.

File: timeline-chart/src/time-graph-container.ts [hunk 1]
```diff
@@ -32,6 +32,16 @@ export class TimeGraphContainer {
...
+
+        let hasCanvas = true;
+
+        if (PIXI.utils.isWebGLSupported()) {
+            if (!canvas.getContext('webgl2')) {
+                hasCanvas = true;
+            } else {
+                hasCanvas = false;
+            }
+        }
```
Comment (mirsky-work) [inline timeline-chart/src/time-graph-container.ts:44]: IMHO, this form and name makes it slightly more readable.

```suggestion
        let noWebgl2 = !PIXI.utils.isWebGLSupported() || !canvas.getContext('webgl2');
```

**Comment by tahini** [comment]
I don't know too much about PIXI. What are the implications of this? Do we still need pixi.js-legacy? Does this improve the performances on machines without webGL?

**Comment by ankusht-work** [comment]
> I don't know too much about PIXI. What are the implications of this? Do we still need pixi.js-legacy? Does this improve the performances on machines without webGL?

Yes, we still need Pixi.js-legacy. The conclusion is that by switching to canvas renderer on machines with no WebGL support we see overall improved performance.

File: timeline-chart/src/time-graph-container.ts (hunk 1 above)
Comment (ankusht-work) [reply timeline-chart/src/time-graph-container.ts]: Updated the commit with suggested changes.

**Review by tahini — APPROVED** [review]
Code looks good. I can't say I see any performance improvement though, my CPU fans still run like mad without webgl2 support. Is that what you observe too?

**Comment by tahini** [comment]
I just saw the improvement with @ankusht-work 's machine. So I totally approve this PR. My CPU fan's problem is the re-rendering of the canvas, which is totally unrelated to this PR.

**Comment by bhufmann** [comment]
> I just saw the improvement with @ankusht-work 's machine. So I totally approve this PR. My CPU fan's problem is the re-rendering of the canvas, which is totally unrelated to this PR.

Thanks for the approval. Yes, it's probably related to the periodic redraw of the canvas. To not periodic redraw the canvas and redraw on demand will be tracked in a different issue.

**Pull request merged**